src/graphforge/executor/executor.py

```
@@ -2951,27 +2956,60 @@ def _execute_union(self, op: Union, input_rows: list[ExecutionContext]) -> list[
+                if isinstance(sub_row, dict):
+                    # Sub_row is a dict from Project - combine with outer bindings
+                    combined_ctx.bindings = {**outer_ctx.bindings, **sub_row}
+                else:
+                    # Sub_row is an ExecutionContext
+                    combined_ctx.bindings = {**outer_ctx.bindings, **sub_row.bindings}
+                result.append(combined_ctx)
+
+        return result
+
     def _execute_subquery(
         self, op: Subquery, input_rows: list[ExecutionContext]
```

Ruff (0.15.1)

[warning] Unused method argument: `op`

(ARG002)